From: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
Date: Wed, 1 Aug 2018 10:33:05 +0200
Subject: [PATCH] sched: drop is_special_task_state() check from
 __set_current_state_no_track()

The is_special_task_state() check in __set_current_state_no_track()
has been wrongly placed. __set_current_state_no_track() is used in RT
while a sleeping lock is acquired. It is used at the begin of the wait
loop with TASK_UNINTERRUPTIBLE and while leaving it and restoring the
original state. The latter part triggers the warning.

Drop the special state check. This is only used within the sleeping lock
implementation and the assignment happens while the PI lock is held.
While at it, drop set_current_state_no_track() because it has no users.

Cc: stable-rt@vger.kernel.org
Signed-off-by: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
---
 include/linux/sched.h |   12 +-----------
 1 file changed, 1 insertion(+), 11 deletions(-)

--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -134,16 +134,7 @@ struct task_group;
 	} while (0)
 
 #define __set_current_state_no_track(state_value)		\
-	do {							\
-		WARN_ON_ONCE(is_special_task_state(state_value));\
-		current->state = (state_value);			\
-	} while (0)
-
-#define set_current_state_no_track(state_value)			\
-	do {							\
-		WARN_ON_ONCE(is_special_task_state(state_value));\
-		smp_store_mb(current->state, (state_value));	\
-	} while (0)
+		current->state = (state_value);
 
 #define set_special_state(state_value)					\
 	do {								\
@@ -199,7 +190,6 @@ struct task_group;
 	smp_store_mb(current->state, (state_value))
 
 #define __set_current_state_no_track(state_value)	__set_current_state(state_value)
-#define set_current_state_no_track(state_value)		set_current_state(state_value)
 
 /*
  * set_special_state() should be used for those states when the blocking task
